File: iosApp/iosApp/Views/Notification/NotificationViewModel.swift
//
//  NotificationViewModel.swift
//  iosApp
//
//  Created by Isabella Aigner on 12.04.23.
//  Copyright © 2023 Ludwig Boltzmann Institute for
//  Digital Health and Prevention - A research institute
//  of the Ludwig Boltzmann Gesellschaft,
//  Oesterreichische Vereinigung zur Foerderung
//  der wissenschaftlichen Forschung
//  Licensed under the Apache 2.0 license with Commons Clause
//  (see https://www.apache.org/licenses/LICENSE-2.0 and
//  https://commonsclause.com/).
//

import Combine
import KMPNativeCoroutinesCombine
import shared

class NotificationViewModel: ObservableObject {
    private let filterViewModel: CoreNotificationFilterViewModel
    private let coreModel: CoreNotificationViewModel

    @Published var notificationList: [NotificationModel] = []

    @Published var filterText: String = ""

    private var cancellables = Set<AnyCancellable>()

    init(filterViewModel: CoreNotificationFilterViewModel) {
        self.filterViewModel = filterViewModel
        coreModel = CoreNotificationViewModel(coreFilterModel: filterViewModel, notificationManager: AppDelegate.shared.notificationManager)

        createPublisher(for: coreModel.notificationList)
        .removeDuplicates()
        .receive(on: DispatchQueue.main)
        .sink(receiveCompletion: { _ in }) { [weak self] notifications in
            self?.notificationList = notifications
        }
        .store(in: &cancellables)

        createPublisher(for: filterViewModel.activeTypes)
            .map { (types: Set<String>) -> String in
                guard !types.isEmpty else {
                    return ""
                }

                let localized = types
                    .sorted()
                    .map { String(localized: String.LocalizationValue($0)) }

                return ListFormatter.localizedString(byJoining: localized)
            }
        .removeDuplicates()
        .receive(on: DispatchQueue.main)
        .sink(receiveCompletion: { _ in }) { [weak self] text in
            self?.filterText = text
        }
        .store(in: &cancellables)
    }

    func handleNotificationAction(notification: NotificationModel, navigationModalState: NavigationModalState) {
        coreModel.handleNotificationAction(notification: notification) { actionHandler, data in
            if let data {
                switch actionHandler {
                case NotificationActionHandler.deeplink:
                    AppDelegate.navigationScreenHandler.openRoute(to: data)
                default:
                    return
                }
            }
        }
    }
}

File: androidApp/src/main/java/io/redlink/umm/blendedcare/app/android/activities/notification/NotificationViewModel.kt
/*
 * Copyright LBI-DHP and/or licensed to LBI-DHP under one or more
 * contributor license agreements (LBI-DHP: Ludwig Boltzmann Institute
 * for Digital Health and Prevention -- A research institute of the
 * Ludwig Boltzmann Gesellschaft, Österreichische Vereinigung zur
 * Förderung der wissenschaftlichen Forschung).
 * Licensed under the Apache 2.0 license with Commons Clause
 * (see https://www.apache.org/licenses/LICENSE-2.0 and
 * https://commonsclause.com/).
 */
package io.redlink.umm.blendedcare.app.android.activities.notification

import androidx.core.net.toUri
import androidx.lifecycle.ViewModel
import androidx.navigation.NavController
import io.redlink.umm.blendedcare.app.android.MoreApplication
import io.redlink.umm.blendedcare.app.android.R
import io.redlink.umm.blendedcare.app.android.extensions.applicationId
import io.redlink.umm.blendedcare.app.android.extensions.stringResource
import io.redlink.umm.participant.models.NotificationModel
import io.redlink.umm.participant.services.notification.NotificationActionHandler
import io.redlink.umm.participant.viewModels.notifications.CoreNotificationFilterViewModel
import io.redlink.umm.participant.viewModels.notifications.CoreNotificationViewModel

class NotificationViewModel(private val coreFilterViewModel: CoreNotificationFilterViewModel) :
    ViewModel() {
    val coreViewModel: CoreNotificationViewModel =
        CoreNotificationViewModel(
            coreFilterViewModel,
            MoreApplication.shared!!.notificationManager,
            stringResource(R.string.app_scheme),
            applicationId
        )

    fun handleNotificationAction(notification: NotificationModel, navController: NavController) {
        coreViewModel.handleNotificationAction(notification) { actionType, data ->
            when (actionType) {
                NotificationActionHandler.DEEPLINK -> navController.navigate(data.toUri())
            }
        }
    }

    fun getFilterString(): String {
        if (!coreFilterViewModel.filterActive()) {
            return stringResource(R.string.more_filter_notification_all)
        }
        return coreFilterViewModel.activeTypes.value.joinToString(", ")
    }
}
